catamandu.md

```diff
@@ -0,0 +1,173 @@
+
+```bash
+catmandu convert MARC --fix 'retain_field(_id)' to YAML < ~/command_line_curriculum/data_files/ebook.mrc
+```
+What is happening here? The MARC file `< ~/command_line_curriculum/data_files/ebook.mrc` contains more than one MARC record. For every MARC record catmandu extracts the `_id` field.
+
+Extracting data out of the MARC record itself is a bit more difficult. MARC is an array-an-array, you need indexes to extract the data. For instance the MARC leader is usually in the first field of a MARC record. Indexing usually begins from 0 for the first field of an array.
+
+```bash
+catmandu convert MARC --fix 'retain_field(record.0)' to YAML < ~/command_line_curriculum/data_files/ebook.mrc
+```
+The leader value itself is the fifth entry in the resulting array. So, we need index 4 to extract it:
+
+```bash
+catmandu convert MARC --fix 'copy_field(record.0.4,leader); retain_field(leader)' to YAML < ~/command_line_curriculum/data_files/ebook.mrc
+```
+We used here a `copy_field` fix to extract the value into a field called leader. The `retain_field` fix is used to keep only this leader field in the result. To process MARC data this way would be very verbose, plus you need to know at which index position the fields are that you are interested in. This is something you usually don’t know.
+
+Catmandu introduces Carsten Klee’s MARCspec to ease the extraction of MARC values out of a record. With the marc_map fix the command above would read:
+
+```bash
+marc_map("LDR",leader)
+retain_field(leader)
+```
+Let's create a file named `myfixes.txt` with the content above, so we can run.
+
+```bash
+catmandu convert MARC --fix myfixes.txt to YAML < ~/command_line_curriculum/data_files/ebook.mrc
+```
+To extract the title fields, the field 245 remember?😉, you can write:
+
+```bash
+marc_map("245",title)
+retain_field(title)
+```
+Or, if you are only interested in the $a subfield you could write:
+
+```bash
+marc_map("245a",title)
+retain_field(title)
+```
+More elaborate mappings are possible which we will see shortly. As a warming up, here is some code to extract all the record identifiers, titles and isbn numbers in a MARC file into a CSV listing (which you can open in ~~Excel~~).
+
+* Step 1, create a fix file myfixes.txt containing:
+
+```bash
+marc_map("245",title)
+marc_map("020a",isbn.$append)
+join_field(isbn,",")
+remove_field(record)
+```
+* Step 2, execute this command:
+
+```bash
+catmandu convert MARC --fix myfixes.txt to CSV < ~/command_line_curriculum/data_files/ebook.mrc
+```
+In the fix above we mapped the 245-field to the title. The ISBN is in the 020-field. Because MARC records can contain one or more 020 fields we created an isbn array using the `isbn.$append` syntax. Next we turned the isbn array back into a comma separated string using the `join_field` fix. As last step we deleted all the fields we didn’t need in the output with the `remove_field` syntax.
```

Comment:
```suggestion
In the fix above we mapped the 245-field to the title. The ISBN is in the 020-field. Because MARC records can contain one or more 020 fields we created an isbn array using the `isbn.$append` syntax. Next we turned the isbn array back into a comma separated string using the `join_field` fix. As a last step we deleted all the fields we didn’t need in the output with the `remove_field` syntax.
```